packages/dds/sequence/src/revertibles.ts

```diff
@@ -0,0 +1,185 @@
+		collection,
+	});
+
+	return revertibles;
+}
+
+export function appendLocalDeleteToRevertibles(
+	string: SharedString,
+	interval: SequenceInterval,
+	revertibles: IntervalRevertible[],
+	collection: IntervalCollection<SequenceInterval>,
+) {
+	// TODO: create localReferences to track positions of deleted interval
+	// just use positions on the interval?
+	revertibles.push({
+		event: IntervalEventType.DELETE,
+		interval,
+		start: interval.start,
+		end: interval.end,
+		collection,
+	});
+
+	return revertibles;
+}
+
+export function appendLocalChangeToRevertibles(
+	string: SharedString,
+	newInterval: SequenceInterval,
+	previousInterval: SequenceInterval,
+	revertibles: IntervalRevertible[],
+	collection: IntervalCollection<SequenceInterval>,
+) {
+	// TODO: create localReferences to track positions of previous interval
+	revertibles.push({
+		event: IntervalEventType.CHANGE,
+		interval: newInterval,
+		start: previousInterval.start,
+		end: previousInterval.end,
+		collection,
+	});
+
+	return revertibles;
+}
+
+export function appendLocalPropertyChangedToRevertibles(
+	interval: SequenceInterval,
+	deltas: PropertySet,
+	revertibles: IntervalRevertible[],
+) {
+	revertibles.push({
+		event: IntervalEventType.PROPERTYCHANGED,
+		interval,
+		propertyDeltas: deltas,
+	});
+
+	return revertibles;
+}
+
+function revertLocalAdd(
+	string: SharedString,
+	revertible: TypedRevertible<typeof IntervalEventType.ADD>,
+) {
+	// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+	const id = revertible.interval.getIntervalId()!;
+	// might be better to assert if id is undefined
+	const label = revertible.interval.properties.label;
+	// should I use revertible.collection here or access from sharedstring?
+	string.getIntervalCollection(label).removeIntervalById(id);
+}
+
+function revertLocalDelete(
+	string: SharedString,
+	revertible: TypedRevertible<typeof IntervalEventType.DELETE>,
+) {
+	const label = revertible.interval.properties.label;
+	const start = string.localReferencePositionToPosition(revertible.start);
+	const end = string.localReferencePositionToPosition(revertible.end);
+	const type = revertible.interval.intervalType;
+	const props = revertible.interval.properties;
+	string.getIntervalCollection(label).add(start, end, type, props);
+}
+
+function revertLocalChange(
+	string: SharedString,
+	revertible: TypedRevertible<typeof IntervalEventType.CHANGE>,
+) {
+	// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+	const id = revertible.interval.getIntervalId()!;
+	// might be better to assert if id is undefined
+	const start = string.localReferencePositionToPosition(revertible.start);
+	const end = string.localReferencePositionToPosition(revertible.end);
+
+	revertible.collection.change(id, start, end);
+}
+
+function revertLocalPropertyChanged(
+	string: SharedString,
+	revertible: TypedRevertible<typeof IntervalEventType.PROPERTYCHANGED>,
+) {
+	const label = revertible.interval.properties.label;
```

Comment:
We could store the interval collection on the revertible instead.

Which makes me wonder if we should always do that. Like what if the interval collection is deleted and then a new one with the same label is created?
_from previous pr_

Reply:
interval collection creation/deletion is implicit--once a collection is materialized in memory, it stays around and is valid to use for as long as its associated sharedString is referenced.

We don't bother including empty interval collections in the snapshot, and we don't issue explicit create or delete operations for them.

So it should generally be fine to keep interval collections by reference wherever you want.